### PR TITLE
[PRC] Adds code for Precision, Recall metrics

### PR DESCRIPTION
**Pull request by voletiv**

### CODE REVIEW

**Comment by mikeroberts3000** [comment]
Hi friends, we needed to compute {FID, KID, precision and recall} for a recent project, and we initially thought we would need to use this repo to compute FID and KID, and a separate repo with a different interface to compute precision and recall. Needless to say, we were delighted to find this pull request, enabling us to do everything we needed from the comfort of torch-fidelity.

We were able to compute precision and recall successfully after correcting a minor typo in this PR. Perhaps my colleague @RachithP can describe the typo in this thread.

@voletiv, if you don't mind me asking, is this implementation basically a wrapper around the authors' original implementation? Or is it a complete reimplementation? Does it reproduce any of the results from the original precision and recall paper? How extensively has it been tested? Should I expect it to produce exactly the same results as the authors' implementation?

File: torch_fidelity/metrics.py [hunk 1]
```diff
@@ -262,6 +277,10 @@ def calculate_metrics(**kwargs):
                 metric_kid = kid_featuresdict_to_metric(featuresdict_1, featuresdict_2, feature_layer_kid, **kwargs)
                 metrics.update(metric_kid)
 
+            if have_prc:
+                metric_prc = prc_featuresdict_to_metric(featuresdict_1, featuresdict_2, feature_layer_prc, **kwargs)
+                metrics.update(metric_pr)
```
Comment (RachithP) [inline torch_fidelity/metrics.py:282]: A minor typo - `metrics.update(metric_pr)` -> `metrics.update(metric_prc)`.

File: torch_fidelity/metric_prc.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        precision, recall = calculate_precision_recall_part(features_1, features_2, NN_k, batch_size)
+    else:
+        precision, recall = calculate_precision_recall_full(features_1, features_2, NN_k, batch_size)
+
+    out = {
+        KEY_METRIC_PRECISION: precision,
+        KEY_METRIC_RECALL: recall,
+    }
+
+    vprint(verbose, f'Precision: {out[KEY_METRIC_PRECISION]}, Recall: {out[KEY_METRIC_PRECISION]}')
```
Comment (RachithP) [inline torch_fidelity/metric_prc.py:84]: A small typo here.  `Recall: {out[KEY_METRIC_PRECISION]}')` -> `Recall: {out[KEY_METRIC_RECALL]}')`.

**Comment by RachithP** [comment]
Hi Friends,
As my colleague @mikeroberts3000 mentioned, we were delighted to see this PR. Thanks for the contribution!

There are a few typos that need to be corrected to make this work. and I've mentioned them above.

Another thing I observed was that when I run `calculate_metrics` with `prc=True` for `input1` with 50k images and `input2` with 50k images, the program crashes. To mitigate this, I had to enable `save_cpu_ram=True` flag. I would also like to point out that for input size of 10k images, there was no crash. The system I ran this on has 32GB of CPU RAM and 8GB  of GPU RAM.

The following produces a crash.
```python
metrics_dict = torch_fidelity.calculate_metrics(input1=input1,
                                                            input2=input2,
                                                            cuda=True,
                                                            fid=True,
                                                            kid=True,
                                                            prc=True,
                                                            verbose=True)
```

**Comment by toshas** [comment]
Thank you for the feedback @mikeroberts3000 and @RachithP, my bad that this great PR from @voletiv is still not merged into master and didn't make into a release yet - I could not allocate a day to double-check that precision and recall computations are according to the initial intent of this project and can be trusted (especially now that torch-fidelity is a dependency of torchmetrics). 

I would greatly appreciate a pair of eyes to peer-review the proposed change and a pair of hands to make it as usable as other metrics. I have merged Vikram's PR into a new branch - now anyone can make their PR's by forking from `precision_and_recall` branch, making necessary changes, and submitting a PR to the same branch, until it is ready for master.

Some action items:
- From the first looks, yes, the memory saving may need some work to avoid crashes with standard 50K dataset sizes;
- Cross-checking with the reference implementations may be done by e.g. submitting a PR to https://github.com/toshas/torch-fidelity-stylegan2-ada (similar to the commit https://github.com/toshas/torch-fidelity-stylegan2-ada/commit/8301266d465600cfa93b1f7c1344414ea67319db 
- It looks like these metrics require implementing and registering a new VGG16 feature extractor - it is already available in sample_similarity_lpips.py, so one can probably either import VGG16features from there, or if there is a cyclic dependency, move this class into a new file;
- Minor fixes from the comments made to the original PR;
- Ambiguity between numerical precision and precision metric, which got introduced into documentation (README.md).

**Pull request merged**